sdk/core/azure-core/src/io/body_stream.cpp

```diff
@@ -82,54 +86,58 @@ int64_t MemoryBodyStream::OnRead(uint8_t* buffer, int64_t count, Context const&
   return copy_length;
 }
 
-#if defined(AZ_PLATFORM_POSIX)
-int64_t FileBodyStream::OnRead(uint8_t* buffer, int64_t count, Azure::Core::Context const& context)
+#if defined(AZ_PLATFORM_WINDOWS)
+#pragma warning(push)
+// warning C4996: 'fopen': This function or variable may be unsafe. Consider using fopen_s instead.
+#pragma warning(disable : 4996)
+#endif
+
+FileBodyStream::FileBodyStream(const std::string& filename) : m_offset(0)
 {
-  (void)context;
-  auto result = pread(
-      this->m_fd,
-      buffer,
-      std::min(count, this->m_length - this->m_offset),
-      this->m_baseOffset + this->m_offset);
+  m_fileStreamHolder = fopen(filename.c_str(), "rb");
+  if (m_fileStreamHolder.GetValue() == nullptr)
+  {
+    throw std::runtime_error("Failed to open file for reading.");
+  }
 
-  if (result < 0)
+#if defined(AZ_PLATFORM_WINDOWS)
+  m_filehandle = (HANDLE)_get_osfhandle(_fileno(m_fileStreamHolder.GetValue()));
+
+  LARGE_INTEGER fileSize;
+  if (!GetFileSizeEx(m_filehandle, &fileSize))
   {
-    throw std::runtime_error("Reading error. (Code Number: " + std::to_string(errno) + ")");
+    throw std::runtime_error("Failed to get size of file.");
   }
+  m_length = fileSize.QuadPart;
 
-  this->m_offset += result;
-  return result;
-}
-#elif defined(AZ_PLATFORM_WINDOWS)
-int64_t FileBodyStream::OnRead(uint8_t* buffer, int64_t count, Azure::Core::Context const& context)
-{
-  (void)context;
-  DWORD numberOfBytesRead;
-  auto o = OVERLAPPED();
-  o.Offset = static_cast<DWORD>(this->m_baseOffset + this->m_offset);
-  o.OffsetHigh = static_cast<DWORD>((this->m_baseOffset + this->m_offset) >> 32);
-
-  auto result = ReadFile(
-      this->m_hFile,
-      buffer,
-      // at most 4Gb to be read
-      static_cast<DWORD>(std::min(
-          static_cast<uint64_t>(0xFFFFFFFFUL),
-          static_cast<uint64_t>(std::min(count, (this->m_length - this->m_offset))))),
-      &numberOfBytesRead,
-      &o);
-
-  if (!result)
+  m_parallelBodyStream = new Internal::ParallelFileBodyStream(m_filehandle, 0, m_length);
+
+#elif defined(AZ_PLATFORM_POSIX)
+  m_fileDescriptor = fileno(m_fileStreamHolder.GetValue());
+
+  struct stat finfo;
+  if (fstat(m_fileDescriptor, &finfo))
   {
-    // Check error. of EOF, return bytes read to EOF
-    auto error = GetLastError();
-    if (error != ERROR_HANDLE_EOF)
-    {
-      throw std::runtime_error("Reading error. (Code Number: " + std::to_string(error) + ")");
-    }
+    throw std::runtime_error("Failed to get size of file.");
   }
+  m_length = finfo.st_size;
```

Comment:
On some 32bit platforms, `st_size` can be a 32 bit integer, so it may not work for file larger than 2G/4G?